### DISCUSSION TITLE
Opinionated commit messages

### DISCUSSION THREAD

**Discussion by stianst** [keycloak/keycloak#9126] [Ideas]
With the move to GH issues it seems people are being very creative in how to link PRs to GH issues. I propose we add a GH action that can enforce a commit message format, where I want to propose the following format:

```
This is a brief summary

If I need it I can add more details, but this is optional and not required

Closes #9999
```

In details:

* Summary
  - Required
  - Starts with upper-case character (A-Z)
  - Max 50 characters
  - Does NOT include GH issue link
* Description
  - Separated from summary with empty line
  - Optional
  - Does NOT include GH issue link
* GH issue link
  - Separated from summary or description with empty line
  - Last part of the commit message
  - Limit the keywords supported by GH

For GH issue link I propose we limit the keywords supported by GH to just:

* Closes #NUM
* Fixes keycloak/REPO/#NUM
* Resolves #NUM, resolves #NUM

I've got a rough PoC for the above in https://regex101.com/r/yAKY1Z/1

It's relatively easy to create a commit with multiple lines:

```
git commit -m "Summary" -m "Optional details" -m "Closes #9999"
```

**Comment by DGuhr** [comment]
I vote against this change. Can you elaborate on the why, please? 

I have doubts that this change will result in upping the bar for new contributors even more, if enforced. Also, call me lazy, but 3 times -m is also weird for me, just to comply to this.

**Reply by andreaTP** [reply]
Just as a technical detail, Git does support out-of-the-box multiline commit messages, e.g.:
```bash
$ git commit -m "More
dquote> lines
dquote> here"
```
```bash
$ git log -n 1

commit 983ec629500235a22fa0fc51a4263c4ca5a0b1cf (HEAD -> try)
Author: andreaTP <andrea.peruffo1982@gmail.com>
Date:   Tue Dec 14 09:04:24 2021 +0000

    More
    lines
    here
```

**Reply by andreaTP** [reply]
In general I agree with @DGuhr that enforcing the commit messages format is going to higher the entrance barrier for new contributors.

I would rather start with documenting the guidelines and suggesting the changes during the review without enforcing the rules, from there we can re-evaluate the policy.

**Reply by andreaTP** [reply]
In case we are going to enforce commit messages anyhow, I think we should try to adopt some widely spread convention such as [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/) to take advantage from all the tooling around it (e.g. [commitlint](https://commitlint.js.org/)), even if it's not optimal, we won't have to maintain code for it.

**Reply by DGuhr** [reply]
Imo it would also be a good first step to add a more direct Template considering the format stian mentions to the gh commit page than what we have now: 

```
<!---
Please read https://github.com/keycloak/keycloak/blob/main/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->
```

I don't want to be hard on this, just saying that there are multiple steps between "everyone writes what they want" and "we enforce a very strict rule and automatically deny if they don't comply". Searching the good middle ground here. :)

**Reply by stianst** [reply]
@DGuhr take a look at open PRs and you'll see why. In addition to that it is time consuming and inefficient for humans (us) to have to ask people to update their PRs/comments, and usually that only happens many hours after the PR was sent, while a bot can tell someone straight away, so they can update it while they're already looking/sending the PR, rather than 2 days later or whatever.

**Reply by mposolda** [reply]
I personally like the format Pedro used:
```
[fixes #9135] - Disables pre-loading offline sessions by default
```
It has single line and at the same time, it provides linking between GH issue and GH PR. And it is more compliant with our existing commit messages based on JIRA like: `KEYCLOAK-12345 Disables pre-loading offline sessions by default`

+1 to enforce the format once we agree how it would look like.

**Comment by vmuzikar** [comment]
If we provide a clear template for PRs, IMHO enforcing it would be useful in the end. In the past, we often needed to explicitly ask contributor for creating a JIRA and putting it properly as a commit message prefix. This now transformed to GH Issues but the requirement still remains – have a formal and precise way to link an issue to PR. GH Actions might help with automating it.

We could even consider enforcing the single commit policy. 😜

**Reply by Pepo48** [reply]
+1.
Moreover, many popular projects already have guidelines and it doesn't seem to discourage new contributors. Angular project [1] is a nice example. Or even Google's Flutter[2] has many contribution guidelines or rules (you need to even complete "Contributors License Agreement" before you send a commit) and despite of that it's one of the most popular projects, when it comes to community contributions.

[1] https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
[2] https://github.com/flutter/flutter/wiki/Tree-hygiene#using-git

**Reply by Pepo48** [reply]
Plus we had a discussion about one and a half year ago about how to make PR review process better and faster. One of the suggestions was to use templates on a multiple levels - not just on the commit message level but on the PR or even the code level too (templates for popular IDEs). So I personally perceive this as a next step.

**Reply by andreaTP** [reply]
> popular projects

I would say that if a project is really popular(e.g. big enough and backed by companies) people will contribute regardless.

In the `akka/akka` project, that does have stars in the same order of magnitude as Keycloak they dropped enforcing standards over the commit messages as it was negatively affecting contributions.

**Reply by Pepo48** [reply]
@andreaTP I can always find a counterexample within the similar stars magnitude, which has, on the other hand, successfully applied the standards/guidelines when it comes to the commit messages format[1]. But that's not the point, right?

The point is that in our case we don't speak about essay but rather about brief summary with optional additional details and GH issue link with the keyword. Basically the same approach that has worked prior to GH Issues.

[1] https://github.com/jaegertracing/jaeger/blob/main/CONTRIBUTING_GUIDELINES.md#making-a-change

**Comment by bs-matil** [comment]
To me its fine to enforce a clear semantic.
May one could consider to have the github action optional in the beginning creating a comment which explains how the commit should look like and at some point it gets mandatory.

**Reply by stianst** [reply]
+1 I don't even think we need to make the bot do anything other than commenting, especially not in this case.

**Comment by stianst** [comment]
I don't understand how a PR template helps at all, as GitHub is smart enough to fill the PR for you based on the commit. Personally I don't use the GH UI to create PRs either, I just create them with "gh pr create -f".

In general I think we should have more GH actions and bots that are validating things upfront as it makes the PR process much more streamlined. It's very costly and time consuming to ask people to update their PRs to do things like squashing, formatting the commit message, creating a GH issue, etc.. 

A bot can comment on that straight away, which lets the contributor update it straight away. While a human will not comment until hours or days later, in turn the contributor won't update it until hours or days on top of that.

**Reply by andreaTP** [reply]
Are you looking for something on those lines: https://github.com/wagoid/commitlint-github-action ?
